docs/design/2020-06-27-non-recursive-common-table-expression.md

```diff
@@ -0,0 +1,143 @@
+
+# Proposal: Non-Recursive Common Table Expression
+
+- Author(s):     [Pingyu](https://github.com/pingyu) (Ping Yu)
+- Last updated:  2020-09-15
+- Discussion at: https://github.com/pingcap/tidb/issues/17472
+
+## Abstract
+
+This proposal proposes to support the __Non-recursive Common Table Expression (CTE)__.
+
+## Background
+
+A common table expression (CTE) is a named temporary result set. It is similar to subroutines in programming languages: they can be referenced from multiple places in a query and may refer to themselves, providing recursive queries [5][6].
+
+E.g.
+```sql
+WITH
+  cte1 AS (SELECT a, b FROM table1),
+  cte2 AS (SELECT c, d FROM table2)
+SELECT b, d FROM cte1 JOIN cte2
+WHERE cte1.a = cte2.c;
+```
+
+Using CTE can achieve these goals [4][7]:
+* Simplifying the query, making it more readable
+* Allowing a recursive entity to be enumerated without the need for recursive-stored procedures (proposed in another [document](https://github.com/pingyu/tidb/blob/executor_recursive_cte/docs/design/2020-08-08-recursive-common-table-expression.md)), to deal with tree or graph data [8].
+* Improve performance in some scenario by evaluating a complex expression only once.
+
+Related issues:
+* [#6824](https://github.com/pingcap/tidb/issues/6824) request to support recursive common table expression
+* [#17916](https://github.com/pingcap/tidb/issues/17916) request to support common table expression to run TPC-DS and other analytic workloads
+* [pingcap/parser#289](https://github.com/pingcap/parser/issues/289) request to support common table expression to be compatible with MySQL 8.0
+* [#18752](https://github.com/pingcap/tidb/issues/18752) request CTE to deal with graph data
+
+## Proposal
+
+We design two strategies for CTE: __Merge__ & __Materialization__:
+
+#### Merge
+The CTE is expanded on where it's referenced:
+```sql
+WITH cte(k, v) AS (SELECT i, j FROM t1)
+SELECT v FROM cte WHERE k > 10
+UNION DISTINCT
+SELECT v FROM cte WHERE k < 100
+```
+After "merged":
+```sql
+SELECT v FROM (SELECT i AS k, j AS v FROM t1) WHERE k > 10
+UNION DISTINCT
+SELECT v FROM (SELECT i AS k, j AS v FROM t1) WHERE k < 100
+```
+
+We will add some rules to rewrite the plan trees, similar to `subquery optimization` [2].
+
+_(`Merge` is also called `Inline` is some other DBMS [4][11])_
+
+#### Materialization
```

Comment:
Could you elaborate more on how to implement this optimization in TiDB? For example:
1. Does it depend on temporary table?
2. How to determine whether to use the merge and the materialization optimizations in the query optimizer?

Reply:
BTW, to quickly support non-recursive CTE (may have perennial performance issues) in the short-term, I think we can focus on how to implement non-recursive CTE with the merge optimization in this proposal. We can discuss how to implement the materialization optimization in the future if we have some idea. A suggested roadmap is:

1. short-term (maybe in TiDB 5.0): non-recursive CTE implemented with merge optimization, no need to support `merge()` or `no_merge()` optimizer hints.
2. long-term (maybe in TiDB 6.0): non-recursive CTE implemented with the materialization and merge, we may need to support `merge()` or `no_merge()` optimizer hints to let the user to further control the optimizer behavior.

Reply:
@SunRunAway @nullnotnil what's your opinion on the short and long term roadmap?